### DISCUSSION TITLE
Overriding decorators for testing proposes

### DISCUSSION THREAD

**Discussion by balexandre** [fastify/fastify#5324] [Q&A]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the issue has not already been raised

### Issue

I'm trying to test a mocked decorator

using `AutoLoad` I'm loading one plug-in `/src/plugins/conversionEngine.ts`  as

```javascript
import type { FastifyInstance } from 'fastify';
import fp from 'fastify-plugin';

import { ConversionEngine } from 'src/engines/conversion';
import { ExchangeService } from 'src/services/exchange';
import { CacheService } from 'src/services/databaseKeyValuePair';

export default fp(async (fastify: FastifyInstance) => {
  const conversionEngine = new ConversionEngine(CacheService, ExchangeService);

  fastify.decorate('conversionEngine', conversionEngine);
});

declare module 'fastify' {
  export interface FastifyInstance {
    conversionEngine: ConversionEngine;
  }
}
```

in my test, I want to inject a mocked database instead of using the `CacheService` instance

I've tried:

```javascript
import fastifyServer, { FastifyInstance } from 'fastify';

import { app } from 'src/app';
import { FASTIFY_OPTIONS } from 'src/constants';
import { ExchangeService } from 'src/services/exchange';
import { ConversionEngine } from 'src/engines/conversion';
import { mockedDatabase } from 'test/mocks/DatabaseService';
...

describe('Convert endpoint', () => {
  let server: FastifyInstance;

  before('setup server', async () => {
    // start server
    server = fastifyServer(FASTIFY_OPTIONS);
    // override decorator
    server.decorate('conversionEngine', new ConversionEngine(mockedDatabase, ExchangeService));
    // setup server instance
    await server.register(app, { test: true });
  });
...
```

for the sake of scope, I have also tried to change the order

```javascript
    // start server
    server = fastifyServer(FASTIFY_OPTIONS);
    // setup server instance
    await server.register(app, { test: true });
    // override decorator
    server.decorate('conversionEngine', new ConversionEngine(mockedDatabase, ExchangeService));
```

but it ends up calling the `CacheService` instance anyway, what is the best approach to test an override of a decorator?

**Comment by metcoder95** [comment]
>I'm converting this to a discussion as it seems is not something where `fastify` requires a fix or so

Have you considered passing the Engine as a property of the plugin options? DI in essence?

**Comment by Eomm** [comment]
I think you could try https://github.com/Eomm/fastify-explorer for this use case

**Comment by matthyk** [comment]
I quickly wrote this plugin that allows to overwrite selected decorators with a desired value in the whole plugin structure. But I have not tested it extensively, so no guarantee that it will work 😄 

```js
const mockPlugin = fp(async function (app, { mocks }) {
    const originalDecorate = app.decorate

    app.decorate = function wrapDecorate(name, fn, dependencies) {
        if (mocks[name]) {
            originalDecorate.call(this, name, mocks[name], dependencies)
            return originalDecorate.call(this, name + 'Original', fn, dependencies)
        } else {
            return originalDecorate.call(this, name, fn, dependencies)
        }
    }
})
```
then before you register the `app` plugin you must register the `mockPlugin` as follows:

```js
import fastifyServer, { FastifyInstance } from 'fastify';

import { app } from 'src/app';
import { FASTIFY_OPTIONS } from 'src/constants';
import { ExchangeService } from 'src/services/exchange';
import { ConversionEngine } from 'src/engines/conversion';
import { mockedDatabase } from 'test/mocks/DatabaseService';
...

describe('Convert endpoint', () => {
  let server: FastifyInstance;

  before('setup server', async () => {
    // start server
    server = fastifyServer(FASTIFY_OPTIONS);

    await server.register(mockPlugin, {
      mocks: {
        conversionEngine: new ConversionEngine(mockedDatabase, ExchangeService)
      }
    })

    // setup server instance
    await server.register(app, { test: true });
  });
... 
```

**Comment by balexandre** [comment]
I chose another approach, the `mockery` approach, so there won't be needed to touch the plugin file itself

just wrote the `test/setup.ts` (setup file that runs before mocha tests) to contain the following:

```javascript
import mockery from 'mockery';

import { mockedDatabase } from 'test/mocks/DatabaseService';

mockery.enable({
  warnOnUnregistered: false,
  warnOnReplace: false,
});

mockery.registerMock('src/services/databaseKeyValuePair', { CacheService: mockedDatabase });
```

the last line above means that in every import that asks for `'src/services/databaseKeyValuePair'` the `mockery` will replace it with the given object, in this case, a `CacheService` object containing the mocked database

this way it does not touch `/src/plugins/conversionEngine.ts` file and makes the whole production code work even without tests

**Reply by matthyk** [reply]
If anyone comes across this discussion again. I have written a plugin that simplifies the mocking of plugins, decorators and hooks https://www.npmjs.com/package/fastify-override
